gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java

```diff
@@ -476,18 +517,189 @@ private void initGamerules() {
+
+        // Tick loop
+        float lastTickTime = Float.MIN_NORMAL;
+
+        while (this.running.get()) {
+            try {
+                // Tick all major subsystems:
+                this.currentTickTime = System.currentTimeMillis();
+                long internalDiffTime = System.nanoTime();
+                this.server.watchdog().add(this.currentTickTime, 30, TimeUnit.SECONDS);
+
+                // Tick networking at every tick
+                PlayerConnection[] connections = this.newConnections.toArray(new PlayerConnection[0]);
+                for (int i = 0, arrayLength = connections.length; i < arrayLength; i++) {
+                    connections[i].update(this.currentTickTime, lastTickTime);
+                }
+                for (io.gomint.server.entity.EntityPlayer player : this.players.keySet()) {
+                    player.connection().update(this.currentTickTime, lastTickTime);
+                }
+
+                this.syncTaskManager.update(this.currentTickTime);
+                this.update(this.currentTickTime, lastTickTime);
+
+                this.server.watchdog().done();
+
+                // Check if we got shutdown
+                if (!this.running.get()) {
+                    break;
+                }
+
+                long startSleep = System.nanoTime();
+                long diff = startSleep - internalDiffTime;
+                boolean warn = false;
+                if (diff <= skipNanos) {
+                    long sleepNeeded = (skipNanos - diff) - this.sleepBalance;
+                    this.sleepBalance = 0;
+
+                    LockSupport.parkNanos(sleepNeeded);
+
+                    long endSleep = System.nanoTime();
+                    long sleptFor = endSleep - startSleep;
+                    diff = skipNanos;
+
+                    if (sleptFor > sleepNeeded) {
+                        this.sleepBalance = sleptFor - sleepNeeded;
+                    }
+                } else {
+                    warn = true;
+                }
+
+                lastTickTime = (float) diff / TimeUnit.SECONDS.toNanos(1);
+                this.tps = (1 / (double) lastTickTime);
+
+                // Due to the fact that we
+                if (this.tps > this.targetTPS()) {
+                    this.tps = this.targetTPS();
+                }
+
+                if (warn) {
+                    this.logger.warn("Running behind: {} / {} tps", this.tps, (1 / (skipNanos / (float) TimeUnit.SECONDS.toNanos(1))));
+                }
+            } catch (Throwable t) {
+                this.logger.error("Error during world tick, shutting down server", t);
+                ReportUploader.create()
+                    .exception(ExceptionUtil.wrap(t))
+                    .tag("world-tick")
+                    .property("world-name", folder())
+                    .includeWorlds()
+                    .upload();
+                this.running.set(false);
+                this.server.shutdown();
```

Comment:
Should one world crashing take down the whole instance? There might be other worlds being fine? Maybe unloading the world as a first mitigation would be better

Reply:
Possible, I just decided to readd that behaviour as thars how it currently works.

Reply:
Behaviour changed